### ISSUE TITLE
standalone lldb build fails with cmake error during configuration

### ISSUE THREAD

**Issue by rayellis4** [llvm/llvm-project#56537]
TL;DR -- trying to build llvm/clang and lldb as separate standalone builds fails with cmake error while configuring lldb build

Detailed steps, following https://lldb.llvm.org/resources/build.html:

Download llvm project source
```
  git -b llvmorg-14.0.6 --depth=1 https://github.com/llvm/llvm-project
```

Building llvm/clang separately
```
  mkdir llvm-build
  cmake -B ~/llvm-project/llvm-build -DLLVM_TARGETS_TO_BUILD=X86 -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_BUILD_TYPE=Release -G "Unix Makefiles" ~/llvm-project/llvm
  cd llvm-build
  make
```
is successful -- no build errors, "Hello World!" sample works, etc.

Then building lldb, leaving llvm/clang build in place
```
  mkdir lldb-build
  cmake -B ~/llvm-project/lldb-build -DLLVM_DIR=~/llvm-project/llvm-build/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release -G "Unix Makefiles" ~/llvm-project/lldb
```
fails with cmake errors, starting with
```
	.
CMake Warning at ~/llvm-project/llvm-build/lib/cmake/llvm/AddLLVM.cmake:1752 (message):
  LLVM_EXTERNAL_LIT set to
  ~/llvm-project/lldb-build/bin/llvm-lit, but the path does
  not exist.
Call Stack (most recent call first):
  ~/llvm-project/llvm-build/lib/cmake/llvm/AddLLVM.cmake:1797 (get_llvm_lit_path)
  ~/llvm-project/llvm-build/lib/cmake/llvm/AddLLVM.cmake:1877 (add_lit_target)
  test/API/CMakeLists.txt:4 (add_lit_testsuites)
	.
```
and working up to
```
	.
CMake Error at ~/llvm-project/llvm-build/lib/cmake/llvm/AddLLVM.cmake:1589 (message):
  make_paths_relative() failed due to error 'No such file or directory', with
  stderr

Call Stack (most recent call first):
  ~/llvm-project/llvm/utils/llvm-lit/CMakeLists.txt:13 (make_paths_relative)
```

Building llvm/clang/lldb together
```
  mkdir full-build
  cmake -B ~/llvm-project/llvm-build -DLLVM_TARGETS_TO_BUILD=X86 -DLLVM_ENABLE_PROJECTS="clang;lldb" -DCMAKE_BUILD_TYPE=Release -G "Unix Makefiles" ~/llvm-project/llvm
  cd full-build
  make
```
works fine.

Not stuck behind this, just thought you'd like to know.

And many thanks for sharing your work, it's a fantastic effort.

**Comment by JDevlieghere** [comment]
The bot [is](https://green.lab.llvm.org/green/view/LLDB/job/lldb-cmake-standalone/) happy, so I tried reproducing with your instructions:

```
$ git clone -b llvmorg-14.0.6 --depth=1 https://github.com/llvm/llvm-project
$ cmake -B /tmp/llvm-build -DLLVM_TARGETS_TO_BUILD=X86 -DLLVM_ENABLE_PROJECTS=clang -DCMAKE_BUILD_TYPE=Release -G "Unix Makefiles" /tmp/llvm-project/llvm/
...
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/llvm-build
$ cd llvm-build && make -j
...
[100%] Linking CXX executable ../../../../bin/clang
[100%] Built target clang
[100%] Linking CXX executable ../../../../bin/c-index-test
[100%] Built target c-index-test
$ cmake -B /tmp/lldb-build/ -DLLVM_DIR=/tmp/llvm-build/lib/cmake/llvm -DCMAKE_BUILD_TYPE=Release -G "Unix Makefiles" /tmp/llvm-project/lldb
...
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/lldb-build
$ cd lldb-build && make -j
...
[100%] Built target lldb
[100%] Linking CXX executable ../../bin/lldb-vscode
[100%] Built target lldb-vscode
```

The warning you're seeing is expected, but I'm not able to reproduce the error. The only difference is that I'm using an absolute path. I know there's an issue when using relative paths (because of the `make_paths_relative` call). I wouldn't expect that to matter though (your shell should expand the  `~` to an absolute path) but it might be worth a shot.

**Comment by rayellis4** [comment]
That's my fault, I wasn't clear: the actual runs I made used absolute paths, I just textually replaced my home directory with ~ to make the bug report a little more compact. I see I also neglected to specify the system, Lubuntu 20.04.  Let me run these once more.

**Comment by rayellis4** [comment]
Same results with clean download, new llvm build. Here is the tail of the output, without my "simplifying" path rewrites:
```
    ...
-- Symbols (liblldb): exporting all symbols from the lldb namespace
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
CMake Warning at /home/ellisr/llvm-project-14.0.6/llvm-build/lib/cmake/llvm/AddLLVM.cmake:1752 (message):
  LLVM_EXTERNAL_LIT set to
  /home/ellisr/llvm-project-14.0.6/lldb-build/bin/llvm-lit, but the path does
  not exist.
Call Stack (most recent call first):
  /home/ellisr/llvm-project-14.0.6/llvm-build/lib/cmake/llvm/AddLLVM.cmake:1797 (get_llvm_lit_path)
  /home/ellisr/llvm-project-14.0.6/llvm-build/lib/cmake/llvm/AddLLVM.cmake:1877 (add_lit_target)
  test/API/CMakeLists.txt:4 (add_lit_testsuites)


CMake Error at /home/ellisr/llvm-project-14.0.6/llvm-build/lib/cmake/llvm/AddLLVM.cmake:1589 (message):
  make_paths_relative() failed due to error 'No such file or directory', with
  stderr

Call Stack (most recent call first):
  /home/ellisr/llvm-project-14.0.6/llvm/utils/llvm-lit/CMakeLists.txt:13 (make_paths_relative)


-- Configuring incomplete, errors occurred!
See also "/home/ellisr/llvm-project-14.0.6/lldb-build/CMakeFiles/CMakeOutput.log".
See also "/home/ellisr/llvm-project-14.0.6/lldb-build/CMakeFiles/CMakeError.log".
```
I hadn't realized the warnings were expected, I'll look more closely at the actual error and let you know if I find anything.

A bit more detail about the system: it's a recently updated Lubuntu 20.04 system (5.4.0-122-generic #138-Ubuntu SMP), gcc/g++ 9.4.0, ld 2.34, cmake 3.23.2. Is there anything more you'd like?

Thanks for your help, and thanks again for sharing LLVM.

  -- Ray

BTW -- The cmake configuration works fine if I disable building the lldb tests (-DLLDB_INCLUDE_TESTS=OFF), but you probably knew that already.

**Comment by rayellis4** [comment]
The proximate cause for the error I'm running into lies in the AddLLVM.cmake function `make_paths_relative,` where the variable `Python3_EXECUTABLE` is not defined before being used in an `execute_process` call. Setting the variable as shown below makes the cmake configuration successful:
```
function(make_paths_relative out_pathlist basedir pathlist)
  # Passing ARG_PATH_VALUES as-is to execute_process() makes cmake strip
  # empty list entries. So escape the ;s in the list and do the splitting
  # ourselves. cmake has no relpath function, so use Python for that.
  string(REPLACE ";" "\\;" pathlist_escaped "${pathlist}")
  set(Python3_EXECUTABLE "/usr/bin/python3")  ### Where should this really be done??
  execute_process(COMMAND "${Python3_EXECUTABLE}" "-c" "\n
import os, sys\n
base = sys.argv[1]
```
The following code at the end of llvm-project/lldb/test/CMakeLists.txt
```
if(LLDB_BUILT_STANDALONE)
  # This has to happen *AFTER* add_lit_testsuite.
  if (EXISTS ${LLVM_MAIN_SRC_DIR}/utils/llvm-lit)
    # LLVM's make_paths_relative uses Python3_EXECUTABLE which isn't set in a
    # standalone LLDB build.
    set(Python3_EXECUTABLE ${Python3_EXECUTABLE})
    add_subdirectory(${LLVM_MAIN_SRC_DIR}/utils/llvm-lit ${CMAKE_CURRENT_BINARY_DIR}/llvm-lit)
  endif()
endif()
```
makes it seem like this was anticipated and corrected, but that it's not working in my setup. Python3 is installed in the normal Ubuntu/Lubuntu manner as /usr/bin/python3.8, symlinked from /usr/bin/python3. Is there some other sort of configuration needed for this to work?

  -- Ray
